malariagen_data/anoph/dipclust.py

```diff
@@ -273,3 +280,418 @@ def _diplotype_pairwise_distances(
+            )
+
+        except ValueError:
+            return None, 0  # No cnv data
+
+        # Reindex to match the order of samples to the dendrogram, and fill
+        # any missing samples with NaN - xarray will do this for us by default.
+        ds_cnv_ordered = ds_cnv.set_index(samples="sample_id").reindex(
+            samples=dendro_sample_id_order,
+        )
+
+        # Get the copy number data to plot.
+        cn_mode = ds_cnv_ordered["CN_mode"].values
+
+        # get labels
+        gene_names = ds_cnv["gene_name"].values
+        gene_ids = ds_cnv["gene_id"].values
+        gene_labels = [
+            n if not pd.isna(n) else gene_ids[i] for i, n in enumerate(gene_names)
+        ]
+
+        # Plot the copy number data.
+        # N.B., here we have to use go.Heatmap directly rather than
+        # px.imshow because the latter fails to incorporate zmin,
+        # zmax and colorscale within the trace data, and so these
+        # then get lost later when we try to combined into a single
+        # figure.
+        trace = go.Heatmap(
+            z=cn_mode,
+            y=gene_labels,
+            zmin=0,
+            zmax=4,
+            colorscale=colorscale,
+            showlegend=None,
+        )
+
+        return trace, ds_cnv.sizes["genes"]
+
+    def _dipclust_concat_subplots(
+        self,
+        figures,
+        width,
+        height,
+        row_heights,
+        region: base_params.regions,
+        n_snps: int,
+        sample_sets: Optional[base_params.sample_sets],
+        sample_query: Optional[base_params.sample_query],
+    ):
+        from plotly.subplots import make_subplots  # type: ignore
+
+        title_lines = []
+        if sample_sets is not None:
+            title_lines.append(f"sample sets: {sample_sets}")
+        if sample_query is not None:
+            title_lines.append(f"sample query: {sample_query}")
+        title_lines.append(f"genomic region: {region} ({n_snps} SNPs)")
+        title = "<br>".join(title_lines)
+
+        # make subplots
+        fig = make_subplots(
+            rows=len(figures),
+            cols=1,
+            shared_xaxes=True,
+            vertical_spacing=0.02,
+            row_heights=row_heights,
+        )
+
+        for i, figure in enumerate(figures):
+            if isinstance(figure, go.Figure):
+                # This is a figure, access the traces within it.
+                for trace in range(len(figure["data"])):
+                    fig.append_trace(figure["data"][trace], row=i + 1, col=1)
+            else:
+                # Assume this is a trace, add directly.
+                fig.append_trace(figure, row=i + 1, col=1)
+
+        fig.update_xaxes(visible=False)
+        fig.update_layout(
+            title=title,
+            width=width,
+            height=height,
+            hovermode="closest",
+            plot_bgcolor="white",
+        )
+
+        # COMMENT: this feels overly complex - can we somehow remove the colorbar earlier,
```

Comment:
Not sure - I did try that originally, but was struggling a lot to do so. Also, I think it requires making Figure objects for the heatmaps first, whereas currently we just make heatmap traces, so it kind of adds more code.

Reply:
haven't got time right now to implement the other stuff but feel free to do so